// @sap/cds/lib/deploy/cfUtil.js

const runCommand = require('./runCommand');
const { nullLogger } = require('./logger');

const POLL_RETRIES = 20;
const POLL_TIMEOUT = 500; //ms, since polling takes 1-2sec, repoll almost immediately

class CfUtil {

    /**
     *
     * @param {*} serviceName
     * @param {*} logger
     * @returns true if service is available at marketplace
     */
    static async hasMarketplaceService(serviceName, logger) {
        try {
            const marketplaceCmd = await runCommand('cf', ['marketplace', '-s', serviceName], logger);
            return marketplaceCmd.code === 0;
        } catch (err) {
            throw this._getError(err);
        }
    }

    /**
     *
     * @param {*} serviceName
     * @param {*} logger
     * @returns service, or undefined
     */
    static async getService(serviceName, logger) {
        try {
            const serviceCmd = await runCommand('cf', ['service', serviceName], logger);

            let serviceInfo;
            if (serviceCmd.code === 0) {
                serviceInfo = {
                    name: serviceName,
                    status: /^\s*status\s*:\s*(.*)$/mi.exec(serviceCmd.stdout)[1],
                    service: /^\s*service\s*:\s*(.*)$/mi.exec(serviceCmd.stdout)[1]
                }
            }

            return serviceInfo;
        } catch (err) {
            throw this._getError(err);
        }
    }

    static async _pollService(serviceName, resolve, reject, logger = nullLogger, counter = 0) {
        if (counter >= POLL_RETRIES) {
            logger.log();
            return reject(new Error(`[cds.deploy] - Aborting polling after ${counter} trials.`));
        }

        const serviceCmd = await runCommand('cf', ['service', serviceName]);
        if (serviceCmd.code !== 0) {
            return reject(new Error(`[cds.deploy] - Getting service info failed with code ${serviceCmd.code}.`));
        }

        if (serviceCmd.stdout.includes('succeeded')) {
            logger.log();
            logger.log(serviceCmd.stdout);

            return resolve();
        }

        if (serviceCmd.stdout.includes('failed')) {
            logger.log();
            logger.log(serviceCmd.stdout);

            return reject(new Error(`[cds.deploy] - Service creation failed.`));
        }

        setTimeout(() => {
            logger.write('.');
            counter = counter + 1;
            CfUtil._pollService(serviceName, resolve, reject, logger, counter);
        }, POLL_TIMEOUT);

        return null; // for sonar
    }

    /**
     * Triggers service creation and resolves once service is created or rejects in case of errors
     *
     * @param {*} service
     * @param {*} plan
     * @param {*} serviceName
     * @param {*} logger, logger or null for no logging
     */
    static async createService(service, plan, serviceName, logger) {
        // cf create-service SERVICE PLAN SERVICE_INSTANCE
        return new Promise(async (resolve, reject) => {
            try {
                const createServiceCmd = await runCommand('cf', ['create-service', service, plan, serviceName], logger);
                if (createServiceCmd) {
                    if (createServiceCmd.code !== 0) {
                        return reject(new Error('[cds.deploy] - Create request failed'));
                    }

                    if (createServiceCmd.stdout.includes('already exists')) {
                        return resolve();
                    }
                }

                return CfUtil._pollService(serviceName, resolve, reject, logger);
            } catch (err) {
                reject(this._getError(err));
            }
        });
    }

    /**
     * Returns the serice key
     *
     * @param {*} serviceInstance
     * @param {*} serviceKeyName
     * @param {*} logger
     * @returns service key
     */
    static async getServiceKey(serviceInstance, serviceKeyName, logger) {
        try {
            const createServiceKeyCmd = await runCommand('cf', ['create-service-key', serviceInstance, serviceKeyName], logger);
            if (createServiceKeyCmd.code === 0) {
                const getServiceKeyCmd = await runCommand('cf', ['service-key', serviceInstance, serviceKeyName], logger);

                const keyJsonStr = getServiceKeyCmd.stdout.match(/\{(.*)\}/s);
                return JSON.parse(keyJsonStr[0]);
            }

            return null;
        } catch (err) {
            throw this._getError(err);
        }
    }

    static _getError(err) {
        if (err.code === 'ENOENT') {
            return new Error(`Command 'cf' not found. Make sure you have the Cloud Foundry command line tool installed.`);
        } else {
            return err;
        }
    }
}

module.exports = CfUtil;

/* eslint no-console: off */
