// @sap/generator-cds/lib/constants.js
const constants = {

    FILE_NAME_CDSRC_JSON: '.cdsrc.json',
    FILE_NAME_PACKAGE_JSON: 'package.json',

    ENV_KEY_CDS_IDE: '_CDS_IDE',
    ENV_KEY_CDS_IDE_WEBIDE: 'webide',
    ENV_KEY_CDS_IDE_ECLIPSE: 'eclipse',

    ENV_VAR_GENERATION_LOG: 'CDS_BUILD_OUTPUTFILE',

    // add new module type here
    MODULE_TYPE_DB: 'db',
    MODULE_TYPE_SRV: 'srv',

    DB_TECHNOLOGY_HANA: 'hana',
    DB_TECHNOLOGY_SQLITE: 'sqlite',

    PRIMARY_DB: 'db',
    HDI_CONTAINER_TYPE: 'com.sap.xs.hdi-container',

    SRV_TECHNOLOGY_JAVA: 'java',
    SRV_TECHNOLOGY_NODEJS: 'nodejs',

    DEFAULT_GEN_FOLDER: 'gen',

    DEFAULT_MODULE_FOLDER: {},

    TASKS: {},

    RESOURCE_TYPE_UAA: 'com.sap.xs.uaa',
    RESOURCE_TYPE_MANAGED_SERVICE: 'org.cloudfoundry.managed-service',
    SERVICE_TYPE_XSUAA: 'xsuaa',

    WINDOWS_FORBIDDEN_FOLDER_NAMES: ['CON', 'PRN', 'AUX', 'NUL',
        'COM1', 'COM2', 'COM3', 'COM4', 'COM5', 'COM6', 'COM7', 'COM8', 'COM9',
        'LPT1', 'LPT2', 'LPT3', 'LPT4', 'LPT5', 'LPT6', 'LPT7', 'LPT8', 'LPT9'],

    ODATA_VERSIONS: {
        DEFAULT: 'v2',
        'odatav2': 'v2',
        'odatav4': 'v4'
    }
};

// add new module type here
constants.MODULE_TYPES = [constants.MODULE_TYPE_DB, constants.MODULE_TYPE_SRV];

// add new module type here
constants.TASKS[constants.MODULE_TYPE_DB] = {};
constants.TASKS[constants.MODULE_TYPE_DB][constants.DB_TECHNOLOGY_HANA] = 'hana';

constants.TASKS[constants.MODULE_TYPE_SRV] = {};
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_JAVA] = 'java-cf';
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_NODEJS] = 'node-cf';

// add new module type here
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_DB] = 'db';
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_SRV] = 'srv';

const regex = {
    REGEX_DB_TECHNOLOGY: `^(${constants.DB_TECHNOLOGY_SQLITE}|${constants.DB_TECHNOLOGY_HANA})?$`,
    REGEX_MODULE_TYPE: `(${constants.MODULE_TYPES.join('|')})`,
    REGEX_ODATA_VERSION: `^(odatav2|odatav4)$`
}

const regex2 = {
    // allow wrod chars, underscore and minus as name characters
    REGEX_MODULE_TYPE_WITH_NAME: `((([\\w-]+):)?${regex.REGEX_MODULE_TYPE})`
}

module.exports = {
    ...constants,
    ...regex,
    ...regex2,
    REGEX_MODULES: `^(${regex2.REGEX_MODULE_TYPE_WITH_NAME}(,${regex2.REGEX_MODULE_TYPE_WITH_NAME})*)?$`,
    REGEX_SRV_TECHNOLOGY: `(${constants.SRV_TECHNOLOGY_NODEJS}|${constants.SRV_TECHNOLOGY_JAVA})`
};
